applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.h

```diff
@@ -0,0 +1,449 @@
+    ) override;
+
+/********************************************************************/
+/*    GetValueOnIntegrationPoints                                   */
+/********************************************************************/
+    /**
+    * @brief Get on rVariable a array_1d Value from the Element Constitutive Law
+    * @param rVariable The variable we want to get
+    * @param rValues The results in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void GetValueOnIntegrationPoints(
+        const Variable<array_1d<double, 3>>& rVariable,
+        std::vector<array_1d<double, 3>>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    /**
+    * @brief Get on rVariable a Vector Value from the Element Constitutive Law
+    * @param rVariable The variable we want to get
+    * @param rValues The results in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void GetValueOnIntegrationPoints(
+        const Variable<Vector>& rVariable,
+        std::vector<Vector>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    /**
+    * @brief Get on rVariable a Matrix Value from the Element Constitutive Law
+    * @param rVariable The variable we want to get
+    * @param rValues The results in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void GetValueOnIntegrationPoints(
+        const Variable<Matrix>& rVariable,
+        std::vector<Matrix>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        std::stringstream buffer;
+        buffer << "\"BaseDiscreteElement\" #" << Id();
+        return buffer.str();
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << "\"BaseDiscreteElement\" #" << Id();
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const {
+        pGetGeometry()->PrintData(rOStream);
+    }
+    ///@}
+
+protected:
+
+    ///@name Protected member Variables
+    ///@{
+    std::vector<ConstitutiveLaw::Pointer> mConstitutiveLawVector; /// The vector containing the constitutive laws
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    /**
+    * @brief It initializes the material
+    */
+    virtual void InitializeMaterial();
+
+    /**
+    * This functions calculates both the RHS and the LHS
+    * @param rLeftHandSideMatrix: The LHS
+    * @param rRightHandSideVector: The RHS
+    * @param rCurrentProcessInfo: The current process info instance
+    * @param CalculateStiffnessMatrixFlag: The flag to set if compute the LHS
+    * @param CalculateResidualVectorFlag: The flag to set if compute the RHS
+    */
+    virtual void CalculateAll(
```

Comment:
If you do it this way then please implement it efficient, i.e. if only the LHS or RHS is required that the other one is not computed